<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.353">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Michael Love">

<title>Genomic ranges: GRanges</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="ranges_files/libs/clipboard/clipboard.min.js"></script>
<script src="ranges_files/libs/quarto-html/quarto.js"></script>
<script src="ranges_files/libs/quarto-html/popper.min.js"></script>
<script src="ranges_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="ranges_files/libs/quarto-html/anchor.min.js"></script>
<link href="ranges_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="ranges_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="ranges_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="ranges_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="ranges_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Genomic ranges: GRanges</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p><a href="http://mikelove.github.io">Michael Love</a> </p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<p>In this lecture note, we demonstrate how to specify regions of the genome (genomic ranges) in R/Bioconductor. We will make use of the <em>GenomicRanges</em> package and the <em>GRanges</em> object defined by this package. The package also provides a number of methods that work on this type of object. Finally we will show some extensions to related packages, and other ways of working with genomic range data.</p>
<p>A <em>GRanges</em> object specifies:</p>
<ol type="1">
<li><p>the basepairs involved in the range (these are integer ranges)</p></li>
<li><p>the strand (+ or -)</p></li>
<li><p>the chromosome</p></li>
<li><p>additional metadata (the “score” for the feature, some statistics, etc.)</p></li>
</ol>
<p>Chromosomes in Bioconductor are called “sequences” and specified with the column <code>seqnames</code>, because sometimes we may want to specify ranges of molecules that are not chromosomes, for example the mitochondrial genome, viral DNA, un-incorporated contigs, etc. Although the package name is <em>GenomicRanges</em>, we don’t specify that the molecule is necessarily DNA, as it could also in theory be RNA or protein sequence within which we are referring to ranges.</p>
<section id="interval-ranges" class="level1">
<h1>Interval ranges</h1>
<p>Let’s start with a set of interval ranges, an <em>IRanges</em> object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GenomicRanges)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>ir <span class="ot">&lt;-</span> <span class="fu">IRanges</span>(<span class="at">start=</span><span class="dv">100</span> <span class="sc">+</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="at">width=</span><span class="dv">10</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>ir</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>IRanges object with 5 ranges and 0 metadata columns:
          start       end     width
      &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
  [1]       101       110        10
  [2]       102       111        10
  [3]       103       112        10
  [4]       104       113        10
  [5]       105       114        10</code></pre>
</div>
</div>
<p>We could also specify the same ranges by setting the <em>end</em>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>ir <span class="ot">&lt;-</span> <span class="fu">IRanges</span>(<span class="at">start=</span><span class="dv">100</span> <span class="sc">+</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="at">end=</span><span class="dv">100</span> <span class="sc">+</span> <span class="dv">10</span><span class="sc">:</span><span class="dv">14</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>ir</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>IRanges object with 5 ranges and 0 metadata columns:
          start       end     width
      &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
  [1]       101       110        10
  [2]       102       111        10
  [3]       103       112        10
  [4]       104       113        10
  [5]       105       114        10</code></pre>
</div>
</div>
<p>We can make a plot of these ranges. It’s a bit different than a typical plot, because I want to point out that we have <em>discrete</em> locations in the genome (so the position isn’t defined over <span class="math inline">\(\mathbb{R}\)</span> but over <span class="math inline">\(\mathbb{N}\)</span>). So I’ve shifted the positions of the labels by 0.5 (see the <code>axis</code> command). Another detail to notice is that the ranges are <em>inclusive</em>. So the first range, from 101 to 110 <em>includes</em> the 110 basepair. So I’ve added +1 when I draw the arrows to show it includes the last basepair.</p>
<p>Note that the convention used in Bioconductor to define interval ranges is not universal. The BED convention is that the range does <strong>not</strong> include the start position of the range. So the interval range written <code>101 110</code> in BED format would not include the 101th basepair. These details matter when we want to refer to specific sequences in the genome, e.g.&nbsp;the transcription start site, or TSS, of a gene.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rafalib)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mar=</span><span class="fu">c</span>(<span class="fl">2.5</span>,<span class="fl">2.5</span>,.<span class="dv">5</span>,.<span class="dv">5</span>))</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="fu">nullplot</span>(<span class="fu">start</span>(<span class="fu">range</span>(ir)), <span class="fu">end</span>(<span class="fu">range</span>(ir))<span class="sc">+</span><span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">6</span>, <span class="at">xaxt=</span><span class="st">"n"</span>)</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v=</span><span class="dv">100</span><span class="sc">:</span><span class="dv">115</span>, <span class="at">col=</span><span class="st">"grey"</span>)</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>(<span class="dv">1</span>, <span class="at">at=</span><span class="dv">100</span><span class="sc">:</span><span class="dv">115</span> <span class="sc">+</span> .<span class="dv">5</span>, <span class="at">labels=</span><span class="dv">100</span><span class="sc">:</span><span class="dv">115</span>) </span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="fu">arrows</span>(<span class="fu">start</span>(ir),<span class="fu">seq_along</span>(ir),<span class="fu">end</span>(ir)<span class="sc">+</span><span class="dv">1</span>,<span class="fu">seq_along</span>(ir),</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>       <span class="at">lwd=</span><span class="dv">3</span>,<span class="at">code=</span><span class="dv">3</span>,<span class="at">angle=</span><span class="dv">90</span>,<span class="at">length=</span>.<span class="dv">05</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="ranges_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="genomic-ranges" class="level1">
<h1>Genomic ranges</h1>
<p>So far we’ve just shown interval ranges. To specify a genomic range, we also will come up with the <em>strand</em> of the ranges (+ or -) and the chromosome.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>st <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"+"</span>,<span class="st">"-"</span>),<span class="fu">c</span>(<span class="dv">3</span>,<span class="dv">2</span>))</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>st</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "+" "+" "+" "-" "-"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>gr <span class="ot">&lt;-</span> <span class="fu">GRanges</span>(<span class="at">seqnames=</span><span class="st">"chr1"</span>, <span class="at">ranges=</span>ir, <span class="at">strand=</span>st)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>gr</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 5 ranges and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]     chr1   101-110      +
  [2]     chr1   102-111      +
  [3]     chr1   103-112      +
  [4]     chr1   104-113      -
  [5]     chr1   105-114      -
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>Note that <em>GRanges</em> objects, when printed, also have a note about the genome. Here it says that we have an unspecified genome, and there is no information about the <code>seqlengths</code>, the lengths of the chromosomes. This information (basically in what universe do the ranges live) is important to track along with the ranges themselves, in case the user asks to, e.g.&nbsp;move all ranges 1 million basepairs to the right. This is not necessarily possible if we have a range that ends less than 1 million basepairs from the end of a chromosome. Or for example, if you want to randomly select features from the genome, or randomly shuffle the locations of features in the genome, we needto know where the component sequences start and end.</p>
<p>So the <em>GRanges</em> object really needs to also contain details about the genome itself.</p>
</section>
<section id="bed-files-and-granges" class="level1">
<h1>BED files and GRanges</h1>
<p>As we previewed above, another common format for representing ranges is the plain-text <a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1">BED format</a>, where the intervals are described in an open-closed format: exclusive of the basepair on the left side, inclusive of the basepair on the right side. As these files are common, we quickly can show the equivalent code for reading/manipulating genomic ranges using <a href="https://bedtools.readthedocs.io/en/latest/">bedtools</a> or in Bioconductor:</p>
<p>First, read in some data from BED file format. Here we have some transcript ranges from Ensembl v86 in the hg38 human genome. We also have a set of three query ranges (note the different strand of these).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>chr1query <span class="ot">&lt;-</span> plyranges<span class="sc">::</span><span class="fu">read_bed</span>(<span class="st">"chr1_query.bed"</span>)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>txp <span class="ot">&lt;-</span> plyranges<span class="sc">::</span><span class="fu">read_bed</span>(<span class="st">"chr1_100Mb_v86_txps.bed"</span>)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="fu">genome</span>(chr1query) <span class="ot">&lt;-</span> <span class="st">"hg38"</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="fu">genome</span>(txp) <span class="ot">&lt;-</span> <span class="st">"hg38"</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>chr1query</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 3 ranges and 2 metadata columns:
      seqnames              ranges strand |        name     score
         &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;numeric&gt;
  [1]     chr1 100500001-100600000      * |         foo        99
  [2]     chr1 101200001-101400000      + |         foo        99
  [3]     chr1 101700001-101900000      - |         foo        99
  -------
  seqinfo: 1 sequence from hg38 genome; no seqlengths</code></pre>
</div>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>txp</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 112 ranges and 2 metadata columns:
        seqnames              ranges strand |            name     score
           &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt; &lt;numeric&gt;
    [1]     chr1  99969789-100026892      + | ENST00000465289         0
    [2]     chr1  99969789-100035637      + | ENST00000533028         0
    [3]     chr1  99969979-100022956      + | ENST00000427993         0
    [4]     chr1  99970436-100026979      + | ENST00000370153         0
    [5]     chr1  99993043-100015341      + | ENST00000422078         0
    ...      ...                 ...    ... .             ...       ...
  [108]     chr1 101803486-101837029      - | ENST00000536598         0
  [109]     chr1 101825090-101910154      - | ENST00000468901         0
  [110]     chr1 101859851-101859957      + | ENST00000516908         0
  [111]     chr1 101882516-101882894      + | ENST00000603614         0
  [112]     chr1 101893105-101894047      + | ENST00000440278         0
  -------
  seqinfo: 1 sequence from hg38 genome; no seqlengths</code></pre>
</div>
</div>
<p>We can see the number of transcripts that overlap our query ranges. We can ask for this number in different ways, by specifying what to do with the strand information: count only matching strand or missing strand, count only matching strand for the features with strand information, or count but we ignore strand altogether.</p>
<p>The <code>overlapsAny</code> function returns a logical vector along the object that is its first argument (here <code>txp</code>).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>txp[ <span class="fu">overlapsAny</span>(txp, chr1query) ] <span class="co"># 19</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 19 ranges and 2 metadata columns:
       seqnames              ranges strand |            name     score
          &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt; &lt;numeric&gt;
   [1]     chr1 100352600-100520277      + | ENST00000336454         0
   [2]     chr1 100538137-100542018      + | ENST00000315033         0
   [3]     chr1 100586649-100587431      - | ENST00000455141         0
   [4]     chr1 101236888-101241518      + | ENST00000305352         0
   [5]     chr1 101236927-101239519      + | ENST00000475821         0
   ...      ...                 ...    ... .             ...       ...
  [15]     chr1 101803486-101846973      - | ENST00000338858         0
  [16]     chr1 101803486-101846973      - | ENST00000465523         0
  [17]     chr1 101803486-101996905      - | ENST00000462354         0
  [18]     chr1 101803486-101837029      - | ENST00000536598         0
  [19]     chr1 101825090-101910154      - | ENST00000468901         0
  -------
  seqinfo: 1 sequence from hg38 genome; no seqlengths</code></pre>
</div>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>txp[ <span class="fu">overlapsAny</span>(txp, chr1query[<span class="dv">2</span><span class="sc">:</span><span class="dv">3</span>]) ] <span class="co"># 16</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 16 ranges and 2 metadata columns:
       seqnames              ranges strand |            name     score
          &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt; &lt;numeric&gt;
   [1]     chr1 101236888-101241518      + | ENST00000305352         0
   [2]     chr1 101236927-101239519      + | ENST00000475821         0
   [3]     chr1 101238090-101239164      + | ENST00000475289         0
   [4]     chr1 101243158-101243749      + | ENST00000561748         0
   [5]     chr1 101256274-101256616      + | ENST00000455561         0
   ...      ...                 ...    ... .             ...       ...
  [12]     chr1 101803486-101846973      - | ENST00000338858         0
  [13]     chr1 101803486-101846973      - | ENST00000465523         0
  [14]     chr1 101803486-101996905      - | ENST00000462354         0
  [15]     chr1 101803486-101837029      - | ENST00000536598         0
  [16]     chr1 101825090-101910154      - | ENST00000468901         0
  -------
  seqinfo: 1 sequence from hg38 genome; no seqlengths</code></pre>
</div>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>txp[ <span class="fu">overlapsAny</span>(txp, chr1query, <span class="at">ignore.strand=</span><span class="cn">TRUE</span>) ] <span class="co"># 26</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 26 ranges and 2 metadata columns:
       seqnames              ranges strand |            name     score
          &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt; &lt;numeric&gt;
   [1]     chr1 100352600-100520277      + | ENST00000336454         0
   [2]     chr1 100538137-100542018      + | ENST00000315033         0
   [3]     chr1 100586649-100587431      - | ENST00000455141         0
   [4]     chr1 101235683-101236528      - | ENST00000432195         0
   [5]     chr1 101236888-101241518      + | ENST00000305352         0
   ...      ...                 ...    ... .             ...       ...
  [22]     chr1 101803486-101837029      - | ENST00000536598         0
  [23]     chr1 101825090-101910154      - | ENST00000468901         0
  [24]     chr1 101859851-101859957      + | ENST00000516908         0
  [25]     chr1 101882516-101882894      + | ENST00000603614         0
  [26]     chr1 101893105-101894047      + | ENST00000440278         0
  -------
  seqinfo: 1 sequence from hg38 genome; no seqlengths</code></pre>
</div>
</div>
<p>We can repeat these overlap calls using <code>bedtools intersect</code> (this is a command line program, not run in R):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bedtools</span> intersect <span class="at">-wa</span> <span class="at">-a</span> chr1_100Mb_v86_txps.bed <span class="at">-b</span> chr1_query.bed <span class="kw">|</span> <span class="fu">head</span> <span class="at">-3</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>chr1    100352599   100520277   ENST00000336454 0   +
chr1    100538136   100542018   ENST00000315033 0   +
chr1    100586648   100587431   ENST00000455141 0   -</code></pre>
</div>
</div>
<p>Let’s count first by ignoring strand. This corresponds to the number above when we set <code>ignore.strand=TRUE</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bedtools</span> intersect <span class="at">-wa</span> <span class="at">-a</span> chr1_100Mb_v86_txps.bed <span class="at">-b</span> chr1_query.bed <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      26</code></pre>
</div>
</div>
<p>In <code>bedtools</code> you use an <code>-s</code> to specify you only want to return overlaps with matching strand (and we drop the unstranded features).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bedtools</span> intersect <span class="at">-s</span> <span class="at">-wa</span> <span class="at">-a</span> chr1_100Mb_v86_txps.bed <span class="at">-b</span> chr1_query.bed <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      16</code></pre>
</div>
</div>
</section>
<section id="granges-range-manipulation" class="level1">
<h1>GRanges range manipulation</h1>
<p>Here we show some more basic manipulations of <em>GRanges</em>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>gr <span class="sc">+</span> <span class="dv">10</span> <span class="co"># growing by 10 on either side</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 5 ranges and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]     chr1    91-120      +
  [2]     chr1    92-121      +
  [3]     chr1    93-122      +
  [4]     chr1    94-123      -
  [5]     chr1    95-124      -
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">shift</span>(gr, <span class="dv">10</span>) <span class="co"># shifting by 10</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 5 ranges and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]     chr1   111-120      +
  [2]     chr1   112-121      +
  [3]     chr1   113-122      +
  [4]     chr1   114-123      -
  [5]     chr1   115-124      -
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">resize</span>(gr, <span class="at">width=</span><span class="dv">1</span>) <span class="co"># 1 bp from the start (+) or end (-)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 5 ranges and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]     chr1       101      +
  [2]     chr1       102      +
  [3]     chr1       103      +
  [4]     chr1       113      -
  [5]     chr1       114      -
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">flank</span>(gr, <span class="at">width=</span><span class="dv">2</span>, <span class="at">both=</span><span class="cn">TRUE</span>) <span class="co"># +/- 2 bp from " or "</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 5 ranges and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]     chr1    99-102      +
  [2]     chr1   100-103      +
  [3]     chr1   101-104      +
  [4]     chr1   112-115      -
  [5]     chr1   113-116      -
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>Note that the last two commands, <code>resize</code> and <code>flank</code>, gave back ranges relative to the <em>start</em> of the ranges with positive strand, and to the <em>end</em> of the ranges with negative strand. This is particular useful for working with genes, in which we often care about the region surrounding the <em>transcription start site</em> or TSS of a gene. The TSS will be the left-most basepair of a gene on the positive strand, and the right-most basepair of a gene on the negative strand.</p>
<p>These functions would usually give us more information, but in this case the information was not specified:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">seqnames</span>(gr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>factor-Rle of length 5 with 1 run
  Lengths:    5
  Values : chr1
Levels(1): chr1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">seqinfo</span>(gr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Seqinfo object with 1 sequence from an unspecified genome; no seqlengths:
  seqnames seqlengths isCircular genome
  chr1             NA         NA   &lt;NA&gt;</code></pre>
</div>
</div>
<p>Instead of looking at a toy <em>GRanges</em>, let’s download some actual genomic ranges, the set of human genes, as annotated by the <a href="http://www.ensembl.org/">Ensembl project</a>, version 86.</p>
<p>This data is easily accessible in Bioconductor using the <em>ensembldb</em> package, and the specific Ensembl database package from Bioconductor.</p>
<p>This says there are about 60 thousand genes (according to Ensembl).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ensembldb)</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(EnsDb.Hsapiens.v86) <span class="co"># this pkg is about 75 Mb</span></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>edb <span class="ot">&lt;-</span> EnsDb.Hsapiens.v86</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> <span class="fu">genes</span>(edb)</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(g)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 63970</code></pre>
</div>
</div>
<p>If you try to print <code>g</code> it will just show you the beginning and end:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>g</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 63970 ranges and 6 metadata columns:
                  seqnames            ranges strand |         gene_id   gene_name
                     &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt; &lt;character&gt;
  ENSG00000223972        1       11869-14409      + | ENSG00000223972     DDX11L1
  ENSG00000227232        1       14404-29570      - | ENSG00000227232      WASH7P
  ENSG00000278267        1       17369-17436      - | ENSG00000278267   MIR6859-1
  ENSG00000243485        1       29554-31109      + | ENSG00000243485   MIR1302-2
  ENSG00000237613        1       34554-36081      - | ENSG00000237613     FAM138A
              ...      ...               ...    ... .             ...         ...
  ENSG00000224240        Y 26549425-26549743      + | ENSG00000224240     CYCSP49
  ENSG00000227629        Y 26586642-26591601      - | ENSG00000227629  SLC25A15P1
  ENSG00000237917        Y 26594851-26634652      - | ENSG00000237917     PARP4P1
  ENSG00000231514        Y 26626520-26627159      - | ENSG00000231514     FAM58CP
  ENSG00000235857        Y 56855244-56855488      + | ENSG00000235857     CTBP2P1
                            gene_biotype seq_coord_system      symbol
                             &lt;character&gt;      &lt;character&gt; &lt;character&gt;
  ENSG00000223972 transcribed_unproces..       chromosome     DDX11L1
  ENSG00000227232 unprocessed_pseudogene       chromosome      WASH7P
  ENSG00000278267                  miRNA       chromosome   MIR6859-1
  ENSG00000243485                lincRNA       chromosome   MIR1302-2
  ENSG00000237613                lincRNA       chromosome     FAM138A
              ...                    ...              ...         ...
  ENSG00000224240   processed_pseudogene       chromosome     CYCSP49
  ENSG00000227629 unprocessed_pseudogene       chromosome  SLC25A15P1
  ENSG00000237917 unprocessed_pseudogene       chromosome     PARP4P1
  ENSG00000231514   processed_pseudogene       chromosome     FAM58CP
  ENSG00000235857   processed_pseudogene       chromosome     CTBP2P1
                                        entrezid
                                          &lt;list&gt;
  ENSG00000223972 100287596,100287102,727856,...
  ENSG00000227232                           &lt;NA&gt;
  ENSG00000278267                      102466751
  ENSG00000243485            105376912,100302278
  ENSG00000237613           654835,645520,641702
              ...                            ...
  ENSG00000224240                           &lt;NA&gt;
  ENSG00000227629                           &lt;NA&gt;
  ENSG00000237917                           &lt;NA&gt;
  ENSG00000231514                           &lt;NA&gt;
  ENSG00000235857                           &lt;NA&gt;
  -------
  seqinfo: 357 sequences (1 circular) from GRCh38 genome</code></pre>
</div>
</div>
<p>You can take a look at a single gene:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>g[<span class="dv">1</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 1 range and 6 metadata columns:
                  seqnames      ranges strand |         gene_id   gene_name           gene_biotype
                     &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt; &lt;character&gt;            &lt;character&gt;
  ENSG00000223972        1 11869-14409      + | ENSG00000223972     DDX11L1 transcribed_unproces..
                  seq_coord_system      symbol                       entrezid
                       &lt;character&gt; &lt;character&gt;                         &lt;list&gt;
  ENSG00000223972       chromosome     DDX11L1 100287596,100287102,727856,...
  -------
  seqinfo: 357 sequences (1 circular) from GRCh38 genome</code></pre>
</div>
</div>
<p>To pull out the gene names you use a <code>$</code>. This would give us a 60 thousand length character vector, so we just show the first 5.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>g<span class="sc">$</span>gene_name[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "DDX11L1"   "WASH7P"    "MIR6859-1" "MIR1302-2" "FAM138A"  </code></pre>
</div>
</div>
<p>We can look at the distribution of genes by chromosome. The <code>seqnames</code> returns a <em>factor-Rle</em> where <em>Rle</em> stands for <em>run length encoding</em>. This is a way to compress a long vector which may contain repeated values in <em>runs</em>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">seqnames</span>(g)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>factor-Rle of length 63970 with 357 runs
  Lengths:                       5264                       2232 ...                        523
  Values : 1                          10                         ...         Y                 
Levels(357): 1 10 11 12 13 14 15 16 17 18 ... LRG_187 LRG_239 LRG_311 LRG_721 LRG_741 LRG_93 MT X Y</code></pre>
</div>
</div>
<p>The object <code>g</code> has information about genes on the standard chromosomes, as well as genes or <em>haplotypes</em> of genes on <em>haplotype chromosomes</em>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">seqlevels</span>(g),<span class="dv">100</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  [1] "1"                          "10"                         "11"                        
  [4] "12"                         "13"                         "14"                        
  [7] "15"                         "16"                         "17"                        
 [10] "18"                         "19"                         "2"                         
 [13] "20"                         "21"                         "22"                        
 [16] "3"                          "4"                          "5"                         
 [19] "6"                          "7"                          "8"                         
 [22] "9"                          "CHR_HG107_PATCH"            "CHR_HG126_PATCH"           
 [25] "CHR_HG1311_PATCH"           "CHR_HG1342_HG2282_PATCH"    "CHR_HG1362_PATCH"          
 [28] "CHR_HG142_HG150_NOVEL_TEST" "CHR_HG151_NOVEL_TEST"       "CHR_HG1651_PATCH"          
 [31] "CHR_HG1832_PATCH"           "CHR_HG2021_PATCH"           "CHR_HG2022_PATCH"          
 [34] "CHR_HG2023_PATCH"           "CHR_HG2030_PATCH"           "CHR_HG2058_PATCH"          
 [37] "CHR_HG2062_PATCH"           "CHR_HG2063_PATCH"           "CHR_HG2066_PATCH"          
 [40] "CHR_HG2072_PATCH"           "CHR_HG2095_PATCH"           "CHR_HG2104_PATCH"          
 [43] "CHR_HG2116_PATCH"           "CHR_HG2128_PATCH"           "CHR_HG2191_PATCH"          
 [46] "CHR_HG2213_PATCH"           "CHR_HG2217_PATCH"           "CHR_HG2232_PATCH"          
 [49] "CHR_HG2233_PATCH"           "CHR_HG2235_PATCH"           "CHR_HG2239_PATCH"          
 [52] "CHR_HG2247_PATCH"           "CHR_HG2249_PATCH"           "CHR_HG2288_HG2289_PATCH"   
 [55] "CHR_HG2290_PATCH"           "CHR_HG2291_PATCH"           "CHR_HG2334_PATCH"          
 [58] "CHR_HG26_PATCH"             "CHR_HG986_PATCH"            "CHR_HSCHR10_1_CTG1"        
 [61] "CHR_HSCHR10_1_CTG2"         "CHR_HSCHR10_1_CTG3"         "CHR_HSCHR10_1_CTG4"        
 [64] "CHR_HSCHR10_1_CTG6"         "CHR_HSCHR11_1_CTG1_2"       "CHR_HSCHR11_1_CTG5"        
 [67] "CHR_HSCHR11_1_CTG6"         "CHR_HSCHR11_1_CTG7"         "CHR_HSCHR11_1_CTG8"        
 [70] "CHR_HSCHR11_2_CTG1"         "CHR_HSCHR11_2_CTG1_1"       "CHR_HSCHR11_3_CTG1"        
 [73] "CHR_HSCHR12_1_CTG1"         "CHR_HSCHR12_1_CTG2_1"       "CHR_HSCHR12_2_CTG1"        
 [76] "CHR_HSCHR12_2_CTG2"         "CHR_HSCHR12_2_CTG2_1"       "CHR_HSCHR12_3_CTG2"        
 [79] "CHR_HSCHR12_3_CTG2_1"       "CHR_HSCHR12_4_CTG2"         "CHR_HSCHR12_4_CTG2_1"      
 [82] "CHR_HSCHR12_5_CTG2"         "CHR_HSCHR12_5_CTG2_1"       "CHR_HSCHR12_6_CTG2_1"      
 [85] "CHR_HSCHR13_1_CTG1"         "CHR_HSCHR13_1_CTG3"         "CHR_HSCHR13_1_CTG5"        
 [88] "CHR_HSCHR13_1_CTG8"         "CHR_HSCHR14_1_CTG1"         "CHR_HSCHR14_2_CTG1"        
 [91] "CHR_HSCHR14_3_CTG1"         "CHR_HSCHR14_7_CTG1"         "CHR_HSCHR15_1_CTG1"        
 [94] "CHR_HSCHR15_1_CTG3"         "CHR_HSCHR15_1_CTG8"         "CHR_HSCHR15_2_CTG3"        
 [97] "CHR_HSCHR15_2_CTG8"         "CHR_HSCHR15_3_CTG3"         "CHR_HSCHR15_3_CTG8"        
[100] "CHR_HSCHR15_4_CTG8"        </code></pre>
</div>
</div>
<p>Let’s subset to just the “standard” chromosomes:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> <span class="fu">keepStandardChromosomes</span>(g, <span class="at">pruning.mode=</span><span class="st">"coarse"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The following functions tell you the names and lengths of chromosomes, as well as the name of the genome (repeated for each chromosome).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="fu">seqlevels</span>(g)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "1"  "10" "11" "12" "13" "14" "15" "16" "17" "18" "19" "2"  "20" "21" "22" "3"  "4"  "5"  "6" 
[20] "7"  "8"  "9"  "MT" "X"  "Y" </code></pre>
</div>
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="fu">seqinfo</span>(g)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Seqinfo object with 25 sequences (1 circular) from GRCh38 genome:
  seqnames seqlengths isCircular genome
  1         248956422      FALSE GRCh38
  10        133797422      FALSE GRCh38
  11        135086622      FALSE GRCh38
  12        133275309      FALSE GRCh38
  13        114364328      FALSE GRCh38
  ...             ...        ...    ...
  8         145138636      FALSE GRCh38
  9         138394717      FALSE GRCh38
  MT            16569       TRUE GRCh38
  X         156040895      FALSE GRCh38
  Y          57227415      FALSE GRCh38</code></pre>
</div>
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="fu">genome</span>(g)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       1       10       11       12       13       14       15       16       17       18       19 
"GRCh38" "GRCh38" "GRCh38" "GRCh38" "GRCh38" "GRCh38" "GRCh38" "GRCh38" "GRCh38" "GRCh38" "GRCh38" 
       2       20       21       22        3        4        5        6        7        8        9 
"GRCh38" "GRCh38" "GRCh38" "GRCh38" "GRCh38" "GRCh38" "GRCh38" "GRCh38" "GRCh38" "GRCh38" "GRCh38" 
      MT        X        Y 
"GRCh38" "GRCh38" "GRCh38" </code></pre>
</div>
</div>
<p>We can get the number of genes on each chromosome:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">seqnames</span>(g))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   1   10   11   12   13   14   15   16   17   18   19    2   20   21   22    3    4    5    6    7 
5264 2232 3289 2969 1316 2244 2171 2538 3039 1181 2962 4024 1403  848 1337 3043 2519 2891 2888 2893 
   8    9   MT    X    Y 
2369 2271   37 2399  523 </code></pre>
</div>
</div>
<p>Note that MT (mitochondrial DNA) has many more genes than the other chromosomes, per basepair:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>tab <span class="ot">&lt;-</span> <span class="fu">table</span>(<span class="fu">seqnames</span>(g))</span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>tab</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   1   10   11   12   13   14   15   16   17   18   19    2   20   21   22    3    4    5    6    7 
5264 2232 3289 2969 1316 2244 2171 2538 3039 1181 2962 4024 1403  848 1337 3043 2519 2891 2888 2893 
   8    9   MT    X    Y 
2369 2271   37 2399  523 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>lens <span class="ot">&lt;-</span> <span class="fu">seqlengths</span>(<span class="fu">seqinfo</span>(g))</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>lens</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        1        10        11        12        13        14        15        16        17        18 
248956422 133797422 135086622 133275309 114364328 107043718 101991189  90338345  83257441  80373285 
       19         2        20        21        22         3         4         5         6         7 
 58617616 242193529  64444167  46709983  50818468 198295559 190214555 181538259 170805979 159345973 
        8         9        MT         X         Y 
145138636 138394717     16569 156040895  57227415 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="fu">all.equal</span>(<span class="fu">names</span>(tab), <span class="fu">names</span>(lens))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(tab<span class="sc">/</span>lens)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="ranges_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>If we remove MT, we can see the range better:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>((tab<span class="sc">/</span>lens)[<span class="sc">-</span><span class="fu">which</span>(<span class="fu">names</span>(tab) <span class="sc">==</span> <span class="st">"MT"</span>)], <span class="at">horiz=</span><span class="cn">TRUE</span>, <span class="at">las=</span><span class="dv">1</span>)</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v=</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>)<span class="sc">*</span><span class="fl">1e-5</span>, <span class="at">col=</span><span class="fu">rgb</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,.<span class="dv">5</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="ranges_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>So roughly 1 gene per 50,000 basepairs, averaging per chromosome.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>((lens<span class="sc">/</span>tab)[<span class="sc">-</span><span class="fu">which</span>(<span class="fu">names</span>(tab) <span class="sc">==</span> <span class="st">"MT"</span>)])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 55799.91</code></pre>
</div>
</div>
</section>
<section id="finding-overlaps" class="level1">
<h1>Finding overlaps</h1>
<p>Now we return to the overlaps functionality that we briefly covered above. There are many ways to find overlaps. I first introduce the <em>base Bioconductor</em> functions for computing overlaps, although later I will show a way to compute overlaps using <code>join</code>-style operations, using a package called <em>plyranges</em>.</p>
<p>One scenario is that we have two sets of genomic ranges and we want to know which ranges from the two sets overlap each other. We might make a <em>query</em> set and ask which genes are overlapping the set in our query.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>query <span class="ot">&lt;-</span> <span class="fu">GRanges</span>(<span class="dv">1</span>, <span class="fu">IRanges</span>((<span class="dv">101</span><span class="sc">:</span><span class="dv">110</span>)<span class="sc">*</span><span class="fl">1e6</span> <span class="sc">+</span> <span class="dv">1</span>,<span class="at">width=</span><span class="fl">1e6</span>))</span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>query</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 10 ranges and 0 metadata columns:
       seqnames              ranges strand
          &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt;
   [1]        1 101000001-102000000      *
   [2]        1 102000001-103000000      *
   [3]        1 103000001-104000000      *
   [4]        1 104000001-105000000      *
   [5]        1 105000001-106000000      *
   [6]        1 106000001-107000000      *
   [7]        1 107000001-108000000      *
   [8]        1 108000001-109000000      *
   [9]        1 109000001-110000000      *
  [10]        1 110000001-111000000      *
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>We can get the numeric indices of the overlaps using <code>findOverlaps</code>. Note that we can turn this resulting object into a data.frame, for further processing.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>fo <span class="ot">&lt;-</span> <span class="fu">findOverlaps</span>(query, g)</span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(fo)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Hits object with 6 hits and 0 metadata columns:
      queryHits subjectHits
      &lt;integer&gt;   &lt;integer&gt;
  [1]         1        2333
  [2]         1        2335
  [3]         1        2336
  [4]         1        2337
  [5]         1        2338
  [6]         1        2339
  -------
  queryLength: 10 / subjectLength: 58650</code></pre>
</div>
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tail</span>(fo)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Hits object with 6 hits and 0 metadata columns:
      queryHits subjectHits
      &lt;integer&gt;   &lt;integer&gt;
  [1]        10        2490
  [2]        10        2491
  [3]        10        2492
  [4]        10        2493
  [5]        10        2494
  [6]        10        2495
  -------
  queryLength: 10 / subjectLength: 58650</code></pre>
</div>
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="co"># to count the overlaps for each range in query:</span></span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">factor</span>(<span class="fu">from</span>(fo), <span class="fu">seq_along</span>(query)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
 1  2  3  4  5  6  7  8  9 10 
17  3 15  2  7  8  6 29 44 36 </code></pre>
</div>
</div>
<p>Equivalently, we can use the <code>countOverlaps</code> function for the last command above:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="fu">countOverlaps</span>(query, g)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 17  3 15  2  7  8  6 29 44 36</code></pre>
</div>
</div>
<p>We can convert the Hits object to a data.frame:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a>fo <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(fo)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can also find out which genes overlap which other genes, by just providing a single set of ranges to <code>findOverlaps</code>. There are <em>many</em> overlapping genes in the human genome, though you should note that we obtain at least two rows in the resulting <em>Hits</em> object for every overlap, and so we can filter this object down by insisting that one index be strictly less than the other.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>fo <span class="ot">&lt;-</span> <span class="fu">findOverlaps</span>(g)</span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(fo)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>SelfHits object with 6 hits and 0 metadata columns:
      queryHits subjectHits
      &lt;integer&gt;   &lt;integer&gt;
  [1]         1           1
  [2]         2           2
  [3]         2           3
  [4]         3           2
  [5]         3           3
  [6]         4           4
  -------
  queryLength: 58650 / subjectLength: 58650</code></pre>
</div>
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a>fo <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(fo)</span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>fo <span class="ot">&lt;-</span> fo[fo[,<span class="dv">1</span>] <span class="sc">&lt;</span> fo[,<span class="dv">2</span>],]</span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(fo)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   queryHits subjectHits
3          2           3
12         9          10
20        15          16
26        19          20
30        21          22
37        26          28</code></pre>
</div>
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(fo)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 12078</code></pre>
</div>
</div>
<p>Earlier, when we defined <em>query</em>, we didn’t consider strand, and so the strand of the ranges was defined as <code>*</code>. This means that we get overlaps of any genes whether + or -. But when we ask whether the genes overlap each other, by default <code>findOverlaps</code> will not count an overlap if the ranges are on different strand. To ask how many genes overlap each other ignoring strand, we need to specify <code>ignore.strand=TRUE</code> (which produces more overlaps).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>fo <span class="ot">&lt;-</span> <span class="fu">findOverlaps</span>(g, <span class="at">ignore.strand=</span><span class="cn">TRUE</span>)</span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a>fo <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(fo)</span>
<span id="cb86-3"><a href="#cb86-3" aria-hidden="true" tabindex="-1"></a>fo <span class="ot">&lt;-</span> fo[fo[,<span class="dv">1</span>] <span class="sc">&lt;</span> fo[,<span class="dv">2</span>],]</span>
<span id="cb86-4"><a href="#cb86-4" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(fo)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 29470</code></pre>
</div>
</div>
<p>If we just wanted to subset to the genes which overlap a given range, we can use <code>overlapsAny</code>, as we saw above:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a>g[ <span class="fu">overlapsAny</span>(g, query[<span class="dv">1</span>]) ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 17 ranges and 6 metadata columns:
                  seqnames              ranges strand |         gene_id     gene_name
                     &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;   &lt;character&gt;
  ENSG00000117543        1 100989623-101026088      - | ENSG00000117543          DPH5
  ENSG00000233184        1 101025878-101087268      + | ENSG00000233184 RP11-421L21.3
  ENSG00000252765        1 101133153-101133339      + | ENSG00000252765      SCARNA16
  ENSG00000271578        1 101190520-101190994      + | ENSG00000271578  RP5-1033H2.1
  ENSG00000225938        1 101235683-101236528      - | ENSG00000225938   RP4-575N6.4
              ...      ...                 ...    ... .             ...           ...
  ENSG00000183298        1 101786340-101787219      - | ENSG00000183298       RPSAP19
  ENSG00000118733        1 101802574-101997030      - | ENSG00000118733         OLFM3
  ENSG00000252717        1 101859851-101859957      + | ENSG00000252717     RNU6-352P
  ENSG00000271277        1 101882516-101882894      + | ENSG00000271277   RP11-82O2.2
  ENSG00000162699        1 101893105-101894047      + | ENSG00000162699      DNAJA1P5
                          gene_biotype seq_coord_system        symbol entrezid
                           &lt;character&gt;      &lt;character&gt;   &lt;character&gt;   &lt;list&gt;
  ENSG00000117543       protein_coding       chromosome          DPH5    51611
  ENSG00000233184            antisense       chromosome RP11-421L21.3     &lt;NA&gt;
  ENSG00000252765               scaRNA       chromosome      SCARNA16     &lt;NA&gt;
  ENSG00000271578 processed_pseudogene       chromosome  RP5-1033H2.1     &lt;NA&gt;
  ENSG00000225938            antisense       chromosome   RP4-575N6.4     &lt;NA&gt;
              ...                  ...              ...           ...      ...
  ENSG00000183298 processed_pseudogene       chromosome       RPSAP19     &lt;NA&gt;
  ENSG00000118733       protein_coding       chromosome         OLFM3   118427
  ENSG00000252717                snRNA       chromosome     RNU6-352P     &lt;NA&gt;
  ENSG00000271277 processed_pseudogene       chromosome   RP11-82O2.2     &lt;NA&gt;
  ENSG00000162699 processed_pseudogene       chromosome      DNAJA1P5     &lt;NA&gt;
  -------
  seqinfo: 25 sequences (1 circular) from GRCh38 genome</code></pre>
</div>
</div>
<p>This is equivalent to the following (while the binary operator however doesn’t allow for arguments):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a>g[ g <span class="sc">%over%</span> query[<span class="dv">1</span>] ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 17 ranges and 6 metadata columns:
                  seqnames              ranges strand |         gene_id     gene_name
                     &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;   &lt;character&gt;
  ENSG00000117543        1 100989623-101026088      - | ENSG00000117543          DPH5
  ENSG00000233184        1 101025878-101087268      + | ENSG00000233184 RP11-421L21.3
  ENSG00000252765        1 101133153-101133339      + | ENSG00000252765      SCARNA16
  ENSG00000271578        1 101190520-101190994      + | ENSG00000271578  RP5-1033H2.1
  ENSG00000225938        1 101235683-101236528      - | ENSG00000225938   RP4-575N6.4
              ...      ...                 ...    ... .             ...           ...
  ENSG00000183298        1 101786340-101787219      - | ENSG00000183298       RPSAP19
  ENSG00000118733        1 101802574-101997030      - | ENSG00000118733         OLFM3
  ENSG00000252717        1 101859851-101859957      + | ENSG00000252717     RNU6-352P
  ENSG00000271277        1 101882516-101882894      + | ENSG00000271277   RP11-82O2.2
  ENSG00000162699        1 101893105-101894047      + | ENSG00000162699      DNAJA1P5
                          gene_biotype seq_coord_system        symbol entrezid
                           &lt;character&gt;      &lt;character&gt;   &lt;character&gt;   &lt;list&gt;
  ENSG00000117543       protein_coding       chromosome          DPH5    51611
  ENSG00000233184            antisense       chromosome RP11-421L21.3     &lt;NA&gt;
  ENSG00000252765               scaRNA       chromosome      SCARNA16     &lt;NA&gt;
  ENSG00000271578 processed_pseudogene       chromosome  RP5-1033H2.1     &lt;NA&gt;
  ENSG00000225938            antisense       chromosome   RP4-575N6.4     &lt;NA&gt;
              ...                  ...              ...           ...      ...
  ENSG00000183298 processed_pseudogene       chromosome       RPSAP19     &lt;NA&gt;
  ENSG00000118733       protein_coding       chromosome         OLFM3   118427
  ENSG00000252717                snRNA       chromosome     RNU6-352P     &lt;NA&gt;
  ENSG00000271277 processed_pseudogene       chromosome   RP11-82O2.2     &lt;NA&gt;
  ENSG00000162699 processed_pseudogene       chromosome      DNAJA1P5     &lt;NA&gt;
  -------
  seqinfo: 25 sequences (1 circular) from GRCh38 genome</code></pre>
</div>
</div>
</section>
<section id="plyranges" class="level1">
<h1>plyranges</h1>
<p>As with <em>SummarizedExperiment</em>, there is a “tidy” way to work with <em>GRanges</em>, called <em>plyranges</em>. For more examples of this type of analysis, take a look at some tutorial content I’ve been working on here:</p>
<p><a href="https://tidyomics.github.io/tidy-ranges-tutorial" class="uri">https://tidyomics.github.io/tidy-ranges-tutorial</a></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(plyranges)</span>
<span id="cb92-2"><a href="#cb92-2" aria-hidden="true" tabindex="-1"></a>g <span class="sc">%&gt;%</span> </span>
<span id="cb92-3"><a href="#cb92-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(gene_biotype) <span class="sc">%&gt;%</span></span>
<span id="cb92-4"><a href="#cb92-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="at">width =</span> <span class="fu">mean</span>(width))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 46 rows and 2 columns
              gene_biotype     width
               &lt;character&gt; &lt;numeric&gt;
1   3prime_overlapping_n..   7001.43
2                antisense  20910.71
3   bidirectional_promot..  91166.25
4                IG_C_gene   2365.29
5          IG_C_pseudogene   1012.44
...                    ...       ...
42  transcribed_unitary_..  50086.47
43  transcribed_unproces..  28365.94
44      unitary_pseudogene  15708.17
45  unprocessed_pseudogene   5417.64
46                vaultRNA    127.00</code></pre>
</div>
</div>
<p>Count overlaps of genes in the query ranges:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a>query <span class="sc">%&gt;%</span></span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">num_genes =</span> <span class="fu">count_overlaps</span>(., g))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 10 ranges and 1 metadata column:
       seqnames              ranges strand | num_genes
          &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;
   [1]        1 101000001-102000000      * |        17
   [2]        1 102000001-103000000      * |         3
   [3]        1 103000001-104000000      * |        15
   [4]        1 104000001-105000000      * |         2
   [5]        1 105000001-106000000      * |         7
   [6]        1 106000001-107000000      * |         8
   [7]        1 107000001-108000000      * |         6
   [8]        1 108000001-109000000      * |        29
   [9]        1 109000001-110000000      * |        44
  [10]        1 110000001-111000000      * |        36
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>We can also do the same types of operations as before, with “base <em>GRanges</em>”:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a>g <span class="sc">%&gt;%</span></span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>) <span class="sc">%&gt;%</span></span>
<span id="cb96-3"><a href="#cb96-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(symbol) <span class="sc">%&gt;%</span></span>
<span id="cb96-4"><a href="#cb96-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">anchor_5p</span>() <span class="sc">%&gt;%</span></span>
<span id="cb96-5"><a href="#cb96-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">width=</span><span class="dv">1</span>) <span class="co"># same as resize(..., width=1)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 5 ranges and 1 metadata column:
                  seqnames    ranges strand |      symbol
                     &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
  ENSG00000223972        1     11869      + |     DDX11L1
  ENSG00000227232        1     29570      - |      WASH7P
  ENSG00000278267        1     17436      - |   MIR6859-1
  ENSG00000243485        1     29554      + |   MIR1302-2
  ENSG00000237613        1     36081      - |     FAM138A
  -------
  seqinfo: 25 sequences (1 circular) from GRCh38 genome</code></pre>
</div>
</div>
<p>Plot genes per chromosome:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tibble)</span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb98-3"><a href="#cb98-3" aria-hidden="true" tabindex="-1"></a>g <span class="sc">%&gt;%</span> </span>
<span id="cb98-4"><a href="#cb98-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(seqnames) <span class="sc">%&gt;%</span></span>
<span id="cb98-5"><a href="#cb98-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="at">num_genes =</span> <span class="fu">n</span>()) <span class="sc">%&gt;%</span></span>
<span id="cb98-6"><a href="#cb98-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span></span>
<span id="cb98-7"><a href="#cb98-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(seqnames, num_genes)) <span class="sc">+</span></span>
<span id="cb98-8"><a href="#cb98-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_col</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="ranges_files/figure-html/unnamed-chunk-36-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.4.1 (2024-06-14)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.4.1

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/New_York
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices datasets  utils     methods   base     

other attached packages:
 [1] ggplot2_3.5.1             tibble_3.2.1              plyranges_1.24.0         
 [4] EnsDb.Hsapiens.v86_2.99.0 ensembldb_2.28.0          AnnotationFilter_1.28.0  
 [7] GenomicFeatures_1.56.0    AnnotationDbi_1.66.0      Biobase_2.64.0           
[10] rafalib_1.0.0             GenomicRanges_1.56.1      GenomeInfoDb_1.40.1      
[13] IRanges_2.38.1            S4Vectors_0.42.1          BiocGenerics_0.50.0      
[16] testthat_3.2.1.1          rmarkdown_2.27            devtools_2.4.5           
[19] usethis_3.0.0            

loaded via a namespace (and not attached):
 [1] DBI_1.2.3                   bitops_1.0-8                remotes_2.5.0              
 [4] rlang_1.1.4                 magrittr_2.0.3              matrixStats_1.3.0          
 [7] compiler_4.4.1              RSQLite_2.3.7               png_0.1-8                  
[10] vctrs_0.6.5                 ProtGenerics_1.36.0         stringr_1.5.1              
[13] profvis_0.3.8               pkgconfig_2.0.3             crayon_1.5.3               
[16] fastmap_1.2.0               XVector_0.44.0              ellipsis_0.3.2             
[19] labeling_0.4.3              utf8_1.2.4                  Rsamtools_2.20.0           
[22] promises_1.3.0              sessioninfo_1.2.2           UCSC.utils_1.0.0           
[25] purrr_1.0.2                 bit_4.0.5                   xfun_0.46                  
[28] zlibbioc_1.50.0             cachem_1.1.0                jsonlite_1.8.8             
[31] blob_1.2.4                  later_1.3.2                 DelayedArray_0.30.1        
[34] BiocParallel_1.38.0         parallel_4.4.1              R6_2.5.1                   
[37] stringi_1.8.4               RColorBrewer_1.1-3          rtracklayer_1.64.0         
[40] pkgload_1.4.0               brio_1.1.5                  Rcpp_1.0.13                
[43] SummarizedExperiment_1.34.0 knitr_1.48                  httpuv_1.6.15              
[46] Matrix_1.7-0                tidyselect_1.2.1            rstudioapi_0.16.0          
[49] abind_1.4-5                 yaml_2.3.10                 codetools_0.2-20           
[52] miniUI_0.1.1.1              curl_5.2.1                  pkgbuild_1.4.4             
[55] lattice_0.22-6              withr_3.0.1                 shiny_1.9.1                
[58] KEGGREST_1.44.1             evaluate_0.24.0             urlchecker_1.0.1           
[61] Biostrings_2.72.1           pillar_1.9.0                MatrixGenerics_1.16.0      
[64] generics_0.1.3              RCurl_1.98-1.16             munsell_0.5.1              
[67] scales_1.3.0                xtable_1.8-4                glue_1.7.0                 
[70] lazyeval_0.2.2              tools_4.4.1                 BiocIO_1.14.0              
[73] GenomicAlignments_1.40.0    fs_1.6.4                    XML_3.99-0.17              
[76] grid_4.4.1                  colorspace_2.1-1            GenomeInfoDbData_1.2.12    
[79] restfulr_0.0.15             cli_3.6.3                   fansi_1.0.6                
[82] S4Arrays_1.4.1              dplyr_1.1.4                 gtable_0.3.5               
[85] digest_0.6.36               SparseArray_1.4.8           farver_2.1.2               
[88] rjson_0.2.21                htmlwidgets_1.6.4           memoise_2.0.1              
[91] htmltools_0.5.8.1           lifecycle_1.0.4             httr_1.4.7                 
[94] mime_0.12                   bit64_4.0.5                </code></pre>
</div>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>