<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Michael Love" />


<title>Introduction to Computational Biology</title>

<script src="index_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="index_files/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="index_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="index_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="index_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="index_files/navigation-1.1/tabsets.js"></script>
<link href="index_files/highlightjs-1.1/default.css" rel="stylesheet" />
<script src="index_files/highlightjs-1.1/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>



<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->






<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Introduction to Computational Biology</h1>
<h4 class="author"><em><a href="http://mikelove.github.io">Michael Love</a></em></h4>

</div>


<div id="course-information" class="section level3">
<h3>Course information</h3>
<p>For official course information see Sakai (syllabus, calendar, paper PDFs, etc). This course makes extensive use of R and assumes basic familiarity with base R (not packages) as a prerequisite. A self-quiz is available <a href="https://github.com/biodatascience/compbio/blob/gh-pages/selfquiz/selfquiz.R">here</a>, with answers provided <a href="https://github.com/biodatascience/compbio/raw/gh-pages/selfquiz/selfquiz_answers.rda">here</a>. You can also find a <a href="https://github.com/biodatascience/compbio_src/blob/master/eda/functions.md">list of base R functions</a> that one should be familiar with.</p>
</div>
<div id="schedule-and-course-notes" class="section level3">
<h3>Schedule and course notes</h3>
<p>For <code>Rmd</code> files, go to the <a href="https://github.com/biodatascience/compbio_src">course repo</a> and navigate the directories, or best of all to clone the repo and navigate within RStudio.</p>
<table>
<thead>
<tr class="header">
<th>Week</th>
<th>Topic</th>
<th>Dir.</th>
<th>HW</th>
<th>HTML</th>
<th>Title</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Aug 23</td>
<td>GitHub</td>
<td></td>
<td></td>
<td><a href="github.html">github</a></td>
<td>RStudio, git, and GitHub</td>
</tr>
<tr class="even">
<td>Aug 28 &amp; 30</td>
<td>R &amp; EDA</td>
<td><code>eda</code></td>
<td><a href="https://github.com/biodatascience/compbio_src/blob/master/eda/EDA_hw.Rmd">üî≠</a></td>
<td><a href="eda/EDA.html">EDA</a></td>
<td>Exploratory data analysis</td>
</tr>
<tr class="odd">
<td></td>
<td></td>
<td></td>
<td></td>
<td><a href="eda/NAs.html">NAs</a></td>
<td>Missing values in R</td>
</tr>
<tr class="even">
<td></td>
<td></td>
<td></td>
<td></td>
<td><a href="eda/brain_RNA.html">brain RNA</a></td>
<td>Exploring brain RNA</td>
</tr>
<tr class="odd">
<td>LbrDay &amp; Sep 6</td>
<td>Bioconductor</td>
<td><code>bioc</code></td>
<td><a href="https://github.com/biodatascience/compbio_src/blob/master/bioc/bioc1_HW.Rmd">‚ô´</a></td>
<td><a href="bioc/objects.html">objects</a></td>
<td>Complex data objects: eSet</td>
</tr>
<tr class="even">
<td></td>
<td></td>
<td></td>
<td></td>
<td><a href="bioc/ranges.html">ranges</a></td>
<td>Genomic ranges</td>
</tr>
<tr class="odd">
<td></td>
<td></td>
<td></td>
<td></td>
<td><a href="bioc/GRL.html">GRL</a></td>
<td>GRangesList: lists of ranges</td>
</tr>
<tr class="even">
<td>Sep 11 &amp; 13</td>
<td>Bioconductor</td>
<td><code>bioc</code></td>
<td><a href="https://github.com/biodatascience/compbio_src/blob/master/bioc/bioc2_HW.Rmd">‚ô¨</a></td>
<td><a href="bioc/SE.html">SE</a></td>
<td>The SummarizedExperiment</td>
</tr>
<tr class="odd">
<td></td>
<td></td>
<td></td>
<td></td>
<td><a href="bioc/anno.html">anno</a></td>
<td>Accessing annotations</td>
</tr>
<tr class="even">
<td></td>
<td></td>
<td></td>
<td></td>
<td><a href="bioc/strings.html">strings</a></td>
<td>Manipulating DNA strings</td>
</tr>
<tr class="odd">
<td>Sep 18 &amp; 20</td>
<td>Distances &amp; norm.</td>
<td><code>dist</code></td>
<td><a href="https://github.com/biodatascience/compbio_src/blob/master/dist/dist1_HW.Rmd">‚öñ</a></td>
<td><a href="dist/distances.html">distances</a></td>
<td>Distances in high dimensions</td>
</tr>
<tr class="even">
<td></td>
<td></td>
<td></td>
<td></td>
<td><a href="dist/hclust.html">hclust</a></td>
<td>Hierarchical clustering</td>
</tr>
<tr class="odd">
<td>Sep 25 &amp; 27</td>
<td>Distances &amp; norm.</td>
<td><code>dist</code></td>
<td><a href="https://github.com/biodatascience/compbio_src/blob/master/dist/dist2_HW.Rmd">‚ö†</a></td>
<td><a href="dist/batch.html">batch</a></td>
<td>Batch effects and sources</td>
</tr>
<tr class="even">
<td></td>
<td></td>
<td></td>
<td></td>
<td><a href="dist/sva.html">sva</a></td>
<td>Surrogate variable analysis</td>
</tr>
<tr class="odd">
<td>Oct 2 &amp; 4</td>
<td>Multiple testing</td>
<td><code>test</code></td>
<td></td>
<td><a href="test/multtest.html">multtest</a></td>
<td>FDR and Benjamini-Hochberg</td>
</tr>
<tr class="even">
<td></td>
<td></td>
<td></td>
<td></td>
<td><a href="test/localfdr.html">localfdr</a></td>
<td>Local false discovery rate</td>
</tr>
<tr class="odd">
<td></td>
<td></td>
<td></td>
<td></td>
<td><a href="test/IDR.html">IDR</a></td>
<td>Irreproducible discovery rate</td>
</tr>
<tr class="even">
<td>Oct 9 &amp; 11</td>
<td><em>Mid-term</em></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="odd">
<td>Oct 16 &amp; 18</td>
<td><em>Fall break</em></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td>Oct 23 &amp; 25</td>
<td>Models and EM</td>
<td><code>model</code></td>
<td><a href="https://github.com/biodatascience/compbio_src/blob/master/model/model_HW.Rmd">‚õ∞</a></td>
<td><a href="model/EM.html">EM</a></td>
<td>Expectation maximization</td>
</tr>
<tr class="odd">
<td></td>
<td></td>
<td></td>
<td></td>
<td><a href="model/motif.html">motif</a></td>
<td>EM for finding DNA motifs</td>
</tr>
<tr class="even">
<td>Oct 30 &amp; Nov 1</td>
<td>Hierarchical models</td>
<td><code>hier</code></td>
<td><a href="https://github.com/biodatascience/compbio_src/blob/master/hier/hier_HW.Rmd">‚ôî</a></td>
<td><a href="hier/hierarchical.html">hierarchical</a></td>
<td>Hierarchical models</td>
</tr>
<tr class="odd">
<td></td>
<td></td>
<td></td>
<td></td>
<td><a href="hier/jamesstein.html">jamesstein</a></td>
<td>James-Stein estimator app</td>
</tr>
<tr class="even">
<td>Nov 6 &amp; 8</td>
<td>Signal processing</td>
<td><code>signal</code></td>
<td><a href="https://github.com/biodatascience/compbio_src/blob/master/signal/signal_HW.Rmd">‚õì</a></td>
<td><a href="signal/hmm.html">hmm</a></td>
<td>Hidden Markov Models</td>
</tr>
<tr class="odd">
<td>Nov 13 &amp; 15</td>
<td>Network analysis</td>
<td><code>net</code></td>
<td></td>
<td><a href="net/network.html">network</a></td>
<td>Network analysis</td>
</tr>
<tr class="even">
<td>Nov 20 &amp; Thxg</td>
<td><em>Projects</em></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="odd">
<td>Nov 27 &amp; 29</td>
<td><em>Projects</em></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td>Dec 4 &amp; 6</td>
<td><em>Projects</em></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
</div>
<div id="reading-list" class="section level3">
<h3>Reading list</h3>
<p>I will fill out the reading list as the course goes on. This is not nearly a complete reading list, but just some of literature I will reference in class.</p>
<ul>
<li><strong>What is the role of the computational biologist / statistician?</strong>
<ul>
<li><a href="http://journals.plos.org/plosbiology/article?id=10.1371/journal.pbio.2002050">All biology is computational biology</a> Florian Markowetz</li>
<li><a href="https://www.stat.berkeley.edu/~sandrine/Docs/TerrySelectedWorksSpringer/Version1/Nolan/Nolan.pdf">Questions, Answers and Statistics</a> Deborah Nolan</li>
<li><a href="https://dl.dropboxusercontent.com/u/23421017/50YearsDataScience.pdf">50 Years of Data Science</a> David Donoho</li>
<li><a href="https://projecteuclid.org/euclid.aoms/1177704711">The Future of Data Analysis</a> John Tukey (this article, discussed by Donoho, is from 1962)</li>
<li><a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004961">Ten Simple Rules for Effective Statistical Practice</a> Kass, Caffo, Davidian, Meng, Yu, and Reid</li>
<li><a href="https://projecteuclid.org/euclid.ss/1009213726">Statistical Modeling: The Two Cultures</a> Leo Breiman</li>
</ul></li>
<li><strong>Exploratory data analysis</strong>
<ul>
<li><a href="http://dplyr.tidyverse.org/">dplyr - A Grammar of Data Manipulation</a></li>
<li><a href="http://ggplot2.org/">ggplot2</a></li>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4547484/">The Genotype-Tissue Expression (GTEx) pilot analysis: Multitissue gene regulation in humans</a></li>
</ul></li>
<li><strong>Bioconductor</strong>
<ul>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4509590/">Orchestrating high-throughput genomic analysis with Bioconductor</a> Huber et al</li>
</ul></li>
<li><strong>Distances and normalization</strong>
<ul>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3218662/">Differential expression analysis for sequence count data</a> Simon Anders and Wolfgang Huber</li>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3880143/">Tackling the widespread and critical impact of batch effects in high-throughput data</a> Leek et al</li>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1994707/">Capturing Heterogeneity in Gene Expression Studies by Surrogate Variable Analysis</a> Jeffrey Leek and John Storey</li>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4404308/">Normalization of RNA-seq data using factor analysis of control genes or samples</a> Risso et al</li>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3398141/">Using probabilistic estimation of expression residuals (PEER) to obtain increased power and interpretability of gene expression analyses</a> Stegle et al</li>
</ul></li>
<li><strong>Multiple testing</strong>
<ul>
<li><a href="http://www.jstor.org/stable/2346101">Controlling the False Discovery Rate: A Practical and Powerful Approach to Multiple Testing</a> Yoav Benjamini and Yosef Hochberg</li>
<li><a href="http://genomics.princeton.edu/storeylab//papers/directfdr.pdf">A direct approach to false discovery rates</a> John Storey</li>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC170937/">Statistical significance for genomewide studies</a> John Storey and Robert Tibshirani</li>
<li><a href="http://www.tandfonline.com/doi/abs/10.1198/016214504000000089">Large-scale simultaneous hypothesis testing</a> Bradley Efron</li>
<li><a href="http://dx.doi.org/10.1198/016214501753382129">Empirical Bayes Analysis of a Microarray Experiment</a> Efron et al</li>
<li><a href="https://projecteuclid.org/euclid.aoas/1318514284">Measuring reproducibility of high-throughput experiments</a> Li et al</li>
</ul></li>
<li><strong>Expectation maximization</strong>
<ul>
<li><a href="https://www.nature.com/nbt/journal/v26/n8/full/nbt1406.html">What is the expectation maximization algorithm?</a> Chuong B Do and Serafim Batzoglou</li>
<li><a href="https://people.csail.mit.edu/rameshvs/content/gmm-em.pdf">Gaussian mixture models and the EM algorithm</a> Ramesh Sridharan</li>
<li><a href="http://cs229.stanford.edu/notes/cs229-notes8.pdf">EM algorithm notes</a> Andrew Ng</li>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1538909/">MEME: discovering and analyzing DNA and protein sequence motifs</a> Bailey et al</li>
</ul></li>
<li><strong>Hierarchical models</strong>
<ul>
<li><a href="http://www.statsci.org/smyth/pubs/ebayes.pdf">Linear models and empirical Bayes methods for assessing differential expression in microarray experiments</a> Gordon Smyth</li>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2904972/">Analyzing ‚Äôomics data using hierarchical models</a> Hongkai Ji and X Shirley Liu</li>
<li><a href="http://statweb.stanford.edu/~ckirby/brad/other/Article1977.pdf">Stein‚Äôs Paradox in Statistics</a> Bradley Efron and Carl Morris</li>
<li><a href="https://www.jstor.org/stable/2284155">Stein‚Äôs estimation rule and its competitors - an empirical Bayes approach</a> Bradley Efron and Carl Morris</li>
</ul></li>
<li><strong>Signal processing</strong>
<ul>
<li><a href="http://ieeexplore.ieee.org/abstract/document/1165342/?reload=true">An Introduction to Hidden Markov Models</a> Lawrence Rabiner and Biing-Hwang Juang</li>
<li><a href="http://www.sciencedirect.com/science/article/pii/S0047259X04000260">Hidden Markov models approach to the analysis of array CGH data</a> Fridlyand et al</li>
</ul></li>
<li><strong>Network analysis</strong>
<ul>
<li><a href="http://dx.doi.org/10.1016/j.molcel.2017.08.006">Static And Dynamic DNA Loops Form AP-1 Bound Activation Hubs During Macrophage Development</a> Phanstiel et al</li>
</ul></li>
</ul>
</div>
<div id="resources" class="section level3">
<h3>Resources</h3>
<ul>
<li><a href="http://genomicsclass.github.io/book/pages/resources.html">Online R Classes and Resources</a></li>
<li>Rafael Irizarry and Michael Love, ‚ÄúData Analysis for the Life Sciences‚Äù <a href="https://leanpub.com/dataanalysisforthelifesciences">Free PDF</a>, <a href="http://genomicsclass.github.io/book/">HTML</a></li>
<li>Kasper Hansen ‚ÄúBioconductor for Genomic Data Science‚Äù <a href="https://kasperdanielhansen.github.io/genbioconductor/">HTML</a></li>
<li>Aaron Quinlan ‚ÄúApplied Computational Genomics‚Äù <a href="https://github.com/quinlan-lab/applied-computational-genomics">Slides</a></li>
<li>Jennifer Bryan et al Stat 545 <a href="http://stat545.com/">HTML</a></li>
</ul>
</div>
<div id="wait-how-come-___-is-missing" class="section level3">
<h3>Wait, how come ___ is missing?</h3>
<p>This is not nearly a complete list of topics in computational biology. The students taking the course are mostly graduate students in biostatistics, who have statistical background but not much exposure to genomic or biological datasets. Classic computational biology topics, such as alignment algorithms or molecular dynamics, are not covered, but instead the focus is on exploring genomic datasets and introducing the key statistical models that flourish in the high throughput setting (normalization, false discovery rate calculation, EM algorithm, hierarchical models, HMM, etc.). The course also focuses on R/Bioconductor, as this is a familiar tool for most of the students, and allows them to jump in to the data analysis. The goal is that exposure to these topics and these datasets will allow them to more effectively read the literature and pursue topics in biology and biomedical research.</p>
<hr />
<p>This page was last updated on 11/07/2017.</p>
<ul>
<li><a href="https://github.com/biodatascience/compbio_src">GitHub repo</a></li>
<li><a href="https://github.com/biodatascience/compbio_src/blob/gh-pages/LICENSE">License</a></li>
</ul>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
