<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Michael Love (he/him)" />


<title>Intro to Computational Biology - UNC BIOS/BCB 784</title>

<script src="index_files/header-attrs-2.27/header-attrs.js"></script>
<script src="index_files/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="index_files/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="index_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="index_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="index_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="index_files/navigation-1.1/tabsets.js"></script>
<link href="index_files/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="index_files/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div id="header">



<h1 class="title toc-ignore">Intro to Computational Biology - UNC BIOS/BCB 784</h1>
<h4 class="author"><a href="http://mikelove.github.io">Michael Love</a> (he/him)</h4>

</div>


<div id="course-information" class="section level3">
<h3>Course information</h3>
<p>BIOS/BCB 784</p>
<p>Fall 2024 – 1:25-2:40 MoWe – <a href="https://sph.unc.edu/room/2306-mcgavran-greenberg-hall/">McGavran 2306</a></p>
<p><a href="BIOS784-Fall2024-Gillings-v7.docx">2024 syllabus</a></p>
<p>This course makes extensive use of R and assumes basic familiarity with base R (not packages) as a prerequisite. A self-quiz is available <a href="https://github.com/biodatascience/compbio/blob/gh-pages/selfquiz/selfquiz.R">here</a>, with answers provided <a href="https://github.com/biodatascience/compbio/raw/gh-pages/selfquiz/selfquiz_answers.rda">here</a>. You can also find a <a href="https://github.com/biodatascience/compbio_src/blob/master/eda/functions.md">list of base R functions</a> that one should be familiar with.</p>
<p><strong>For 2024 BCB students</strong>: BCB 720 (background on statistical inference, i.e. working with the likelihood for parameter inference, conditional probabilities) is suitable as a pre-requisite for BIOS/BCB 784.</p>
</div>
<div id="schedule-and-course-notes" class="section level3">
<h3>Schedule and course notes</h3>
<p>For <code>Rmd</code> or <code>qmd</code> files, go to the <a href="https://github.com/biodatascience/compbio_src">course repo</a> and navigate the directories, or clone the repo and navigate within RStudio.</p>
<table style="width:100%;">
<colgroup>
<col width="16%" />
<col width="19%" />
<col width="16%" />
<col width="11%" />
<col width="16%" />
<col width="19%" />
</colgroup>
<thead>
<tr class="header">
<th align="left">Week</th>
<th align="left">Topic</th>
<th align="left">Dir.</th>
<th align="left">HW</th>
<th align="left">HTML</th>
<th align="left">Title</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left"></td>
<td align="left"><strong>Data analysis</strong></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left">Aug 19</td>
<td align="left">Biological intro / GitHub</td>
<td align="left"><code>-</code></td>
<td align="left"><a href="https://github.com/biodatascience/compbio_src/blob/master/github/github_HW.Rmd">HW0</a></td>
<td align="left"><a href="github.html">github</a></td>
<td align="left">RStudio, git, and GitHub</td>
</tr>
<tr class="odd">
<td align="left">Aug 21</td>
<td align="left">Exploring bio data</td>
<td align="left"><code>eda</code></td>
<td align="left"></td>
<td align="left"><a href="eda/EDA.html">EDA</a></td>
<td align="left">Exploratory data analysis</td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"><a href="eda/brain_RNA.html">brain RNA</a></td>
<td align="left">Exploring brain RNA</td>
</tr>
<tr class="odd">
<td align="left">Aug 26 &amp; 28</td>
<td align="left">R/Bioconductor I</td>
<td align="left"><code>bioc</code></td>
<td align="left"><a href="https://github.com/biodatascience/compbio_src/blob/master/bioc/bioc_objects_ranges_HW.Rmd">HW1</a></td>
<td align="left"><a href="bioc/objects.html">objects</a></td>
<td align="left">Bioc data objects</td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"><a href="bioc/ranges.html">ranges</a></td>
<td align="left">Genomic ranges</td>
</tr>
<tr class="odd">
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"><a href="bioc/GRL.html">GRL</a></td>
<td align="left">GRangesList: lists of ranges</td>
</tr>
<tr class="even">
<td align="left">Sep 2</td>
<td align="left">Labor day</td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left">Sep 4</td>
<td align="left">“tidyomics”</td>
<td align="left"><code>TO@GH</code></td>
<td align="left"></td>
<td align="left"><a href="https://tidyomics.github.io/tidy-intro-talk/">tidy intro</a></td>
<td align="left">Tidiness-in-Bioconductor intro</td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"><a href="https://tidyomics.github.io/tidy-ranges-tutorial/">tidy ranges</a></td>
<td align="left">Tidy ranges tutorial</td>
</tr>
<tr class="odd">
<td align="left">Sep 9 &amp; 11</td>
<td align="left">R/Bioconductor II</td>
<td align="left"><code>bioc</code></td>
<td align="left"><a href="https://github.com/biodatascience/compbio_src/blob/master/bioc/bioc_anno_strings_HW.Rmd">HW2</a></td>
<td align="left"><a href="bioc/anno.html">anno</a></td>
<td align="left">Accessing annotations</td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"><a href="bioc/strings.html">strings</a></td>
<td align="left">Manipulating DNA strings</td>
</tr>
<tr class="odd">
<td align="left">Sep 16 &amp; 18</td>
<td align="left">Distances &amp; norm. I</td>
<td align="left"><code>dist</code></td>
<td align="left"><a href="https://github.com/biodatascience/compbio_src/blob/master/dist/dist_biases_scaling_HW.Rmd">HW3</a></td>
<td align="left"><a href="dist/distances.html">distances</a></td>
<td align="left">Distances in high dimensions</td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"><a href="dist/transform_clust.html">transform_clust</a></td>
<td align="left">Transformations and clustering</td>
</tr>
<tr class="odd">
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"><a href="https://github.com/biodatascience/compbio_src/blob/master/dist/vst_math.qmd">vst_math</a></td>
<td align="left">VST math</td>
</tr>
<tr class="even">
<td align="left">Sep 23</td>
<td align="left">Wellness day</td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left">Sep 25</td>
<td align="left">Spatial data analysis (guest lecture)</td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"><a href="https://www.linkedin.com/in/peyton-kuhlers/">Peyton Kuhlers</a> (Hoadley &amp; Raab labs, UNC)</td>
</tr>
<tr class="even">
<td align="left">Sep 30</td>
<td align="left">BCB retreat</td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left">Oct 2</td>
<td align="left">Distances &amp; norm. II</td>
<td align="left"><code>dist</code></td>
<td align="left"></td>
<td align="left"><a href="dist/batch.html">batch</a></td>
<td align="left">Batch effects and sources</td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"><a href="dist/sva.html">sva</a></td>
<td align="left">Surrogate variable analysis</td>
</tr>
<tr class="odd">
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left">Batch effect solutions</td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left"><strong>Data modeling</strong></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left">Oct 7 &amp; 9</td>
<td align="left">Models and EM</td>
<td align="left"><code>model</code></td>
<td align="left"></td>
<td align="left"><a href="model/EM.html">EM</a></td>
<td align="left">Expectation maximization</td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"><a href="model/motif.html">motif</a></td>
<td align="left">EM for finding DNA motifs</td>
</tr>
<tr class="odd">
<td align="left"></td>
<td align="left">ChIP-seq</td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left">Motifs part II</td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left">(EM notes posted to GH)</td>
</tr>
<tr class="odd">
<td align="left">Oct 14 &amp; 16</td>
<td align="left">Hierarchical models</td>
<td align="left"><code>hier</code></td>
<td align="left"></td>
<td align="left"><a href="hier/hierarchical.html">hierarchical</a></td>
<td align="left">Hierarchical models</td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"><a href="hier/jamesstein.html">jamesstein</a></td>
<td align="left">James-Stein estimator</td>
</tr>
<tr class="odd">
<td align="left">Oct 21 &amp; 23</td>
<td align="left">Markov models</td>
<td align="left"><code>markov</code></td>
<td align="left"></td>
<td align="left"><a href="markov/hmm.html">hmm</a></td>
<td align="left">Hidden Markov Models</td>
</tr>
<tr class="even">
<td align="left">Oct 28 &amp; 30</td>
<td align="left">Simultaneous Hypothesis Testing</td>
<td align="left"><code>multiple</code></td>
<td align="left"></td>
<td align="left"><a href="multiple/multtest.html">multtest</a></td>
<td align="left">FDR and Benjamini-Hochberg</td>
</tr>
<tr class="odd">
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"><a href="multiple/localfdr.html">localfdr</a></td>
<td align="left">Local false discovery rate</td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"><a href="multiple/IDR.html">IDR</a></td>
<td align="left">Irreproducible discovery rate</td>
</tr>
<tr class="odd">
<td align="left">Nov 4</td>
<td align="left">TBA</td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left">Nov 6</td>
<td align="left">Guest lecture TBA</td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left">Nov 11 &amp; 23</td>
<td align="left">Network analysis</td>
<td align="left"><code>net</code></td>
<td align="left"></td>
<td align="left"><a href="net/network.html">network</a></td>
<td align="left">Network analysis</td>
</tr>
<tr class="even">
<td align="left">Nov 18 &amp; 20</td>
<td align="left">python for bio data</td>
<td align="left"><code>python</code></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left">Nov 25</td>
<td align="left">Final project presentations</td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left">Dec 2 &amp; 4</td>
<td align="left">Final project presentations</td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
</tr>
</tbody>
</table>
</div>
<div id="reading-list" class="section level3">
<h3>Reading list</h3>
<ul>
<li><strong>What is the role of the computational biologist / statistician?</strong>
<ul>
<li><a href="http://journals.plos.org/plosbiology/article?id=10.1371/journal.pbio.2002050">All biology is computational biology</a> Florian Markowetz</li>
<li><a href="https://www.stat.berkeley.edu/~sandrine/Docs/TerrySelectedWorksSpringer/Version1/Nolan/Nolan.pdf">Questions, Answers and Statistics</a> Deborah Nolan</li>
<li><a href="https://dl.dropboxusercontent.com/u/23421017/50YearsDataScience.pdf">50 Years of Data Science</a> David Donoho</li>
<li><a href="https://projecteuclid.org/euclid.aoms/1177704711">The Future of Data Analysis</a> John Tukey (this article, discussed by Donoho, is from 1962)</li>
<li><a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004961">Ten Simple Rules for Effective Statistical Practice</a> Kass, Caffo, Davidian, Meng, Yu, and Reid</li>
<li><a href="https://projecteuclid.org/euclid.ss/1009213726">Statistical Modeling: The Two Cultures</a> Leo Breiman</li>
</ul></li>
<li><strong>Exploratory data analysis</strong>
<ul>
<li><a href="http://dplyr.tidyverse.org/">dplyr - A Grammar of Data Manipulation</a></li>
<li><a href="http://ggplot2.org/">ggplot2</a></li>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4547484/">The Genotype-Tissue Expression (GTEx) pilot analysis: Multitissue gene regulation in humans</a></li>
</ul></li>
<li><strong>Bioconductor</strong>
<ul>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4509590/">Orchestrating high-throughput genomic analysis with Bioconductor</a> Huber et al</li>
</ul></li>
<li><strong>Distances and normalization</strong>
<ul>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3218662/">Differential expression analysis for sequence count data</a> Simon Anders and Wolfgang Huber</li>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3880143/">Tackling the widespread and critical impact of batch effects in high-throughput data</a> Leek et al</li>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1994707/">Capturing Heterogeneity in Gene Expression Studies by Surrogate Variable Analysis</a> Jeffrey Leek and John Storey</li>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4404308/">Normalization of RNA-seq data using factor analysis of control genes or samples</a> Risso et al</li>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3398141/">Using probabilistic estimation of expression residuals (PEER) to obtain increased power and interpretability of gene expression analyses</a> Stegle et al</li>
</ul></li>
<li><strong>Multiple testing</strong>
<ul>
<li><a href="http://www.jstor.org/stable/2346101">Controlling the False Discovery Rate: A Practical and Powerful Approach to Multiple Testing</a> Yoav Benjamini and Yosef Hochberg</li>
<li><a href="http://genomics.princeton.edu/storeylab//papers/directfdr.pdf">A direct approach to false discovery rates</a> John Storey</li>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC170937/">Statistical significance for genomewide studies</a> John Storey and Robert Tibshirani</li>
<li><a href="http://www.tandfonline.com/doi/abs/10.1198/016214504000000089">Large-scale simultaneous hypothesis testing</a> Bradley Efron</li>
<li><a href="http://dx.doi.org/10.1198/016214501753382129">Empirical Bayes Analysis of a Microarray Experiment</a> Efron et al</li>
<li><a href="https://projecteuclid.org/euclid.aoas/1318514284">Measuring reproducibility of high-throughput experiments</a> Li et al</li>
</ul></li>
<li><strong>Expectation maximization</strong>
<ul>
<li><a href="https://www.nature.com/nbt/journal/v26/n8/full/nbt1406.html">What is the expectation maximization algorithm?</a> Chuong B Do and Serafim Batzoglou</li>
<li><a href="https://people.csail.mit.edu/rameshvs/content/gmm-em.pdf">Gaussian mixture models and the EM algorithm</a> Ramesh Sridharan</li>
<li><a href="http://cs229.stanford.edu/notes/cs229-notes8.pdf">EM algorithm notes</a> Andrew Ng</li>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1538909/">MEME: discovering and analyzing DNA and protein sequence motifs</a> Bailey et al</li>
</ul></li>
<li><strong>Hierarchical models</strong>
<ul>
<li><a href="http://www.statsci.org/smyth/pubs/ebayes.pdf">Linear models and empirical Bayes methods for assessing differential expression in microarray experiments</a> Gordon Smyth</li>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2904972/">Analyzing ’omics data using hierarchical models</a> Hongkai Ji and X Shirley Liu</li>
<li><a href="http://statweb.stanford.edu/~ckirby/brad/other/Article1977.pdf">Stein’s Paradox in Statistics</a> Bradley Efron and Carl Morris</li>
<li><a href="https://www.jstor.org/stable/2284155">Stein’s estimation rule and its competitors - an empirical Bayes approach</a> Bradley Efron and Carl Morris</li>
</ul></li>
<li><strong>Signal processing</strong>
<ul>
<li><a href="http://ieeexplore.ieee.org/abstract/document/1165342/?reload=true">An Introduction to Hidden Markov Models</a> Lawrence Rabiner and Biing-Hwang Juang</li>
<li><a href="http://www.sciencedirect.com/science/article/pii/S0047259X04000260">Hidden Markov models approach to the analysis of array CGH data</a> Fridlyand et al</li>
</ul></li>
<li><strong>Network analysis</strong>
<ul>
<li><a href="http://dx.doi.org/10.1016/j.molcel.2017.08.006">Static And Dynamic DNA Loops Form AP-1 Bound Activation Hubs During Macrophage Development</a> Phanstiel et al</li>
</ul></li>
</ul>
</div>
<div id="resources" class="section level3">
<h3>Resources</h3>
<ul>
<li><a href="http://genomicsclass.github.io/book/pages/resources.html">Online R Classes and Resources</a></li>
<li>Rafael Irizarry and Michael Love, “Data Analysis for the Life Sciences” <a href="https://leanpub.com/dataanalysisforthelifesciences">Free PDF</a>, <a href="http://genomicsclass.github.io/book/">HTML</a></li>
<li><a href="https://kasperdanielhansen.github.io/genbioconductor/">Kasper Hansen, “Bioconductor for Genomic Data Science”</a></li>
<li><a href="https://github.com/quinlan-lab/applied-computational-genomics">Aaron Quinlan, “Applied Computational Genomics” (Slides)</a></li>
<li><a href="http://stat545.com/">Jennifer Bryan et al, Stat 545</a></li>
<li><a href="https://doi.org/10.1371/journal.pcbi.1004387">Florian Markowetz, “You Are Not Working for Me; I Am Working with You”</a></li>
<li><a href="tips.html">Tips to succeed in Computational Biology research</a></li>
</ul>
<p>Some R resources</p>
<ul>
<li><a href="http://www.statmethods.net/">Quick-R</a></li>
<li><a href="http://mathesaurus.sourceforge.net/octave-r.html">R for Matlab users</a></li>
<li><a href="http://www.cookbook-r.com/">R Cookbook</a></li>
<li><a href="https://cran.r-project.org/doc/contrib/Short-refcard.pdf">R Reference Card</a></li>
<li><a href="http://adv-r.had.co.nz/">Advanced R</a></li>
<li><a href="https://github.com/mikelove/bioc-refcard">Bioconductor Reference Card</a></li>
</ul>
<hr />
<p>This page was last updated on 09/18/2024.</p>
<ul>
<li><a href="https://github.com/biodatascience/compbio_src">GitHub repo</a></li>
<li><a href="https://github.com/biodatascience/compbio_src/blob/master/LICENSE">License</a></li>
</ul>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
