---
title: "Introduction to Computational Biology - UNC BIOS / BCB 784"
author: "[Michael Love](http://mikelove.github.io) (he/him)"
output: 
  html_document:
    self_contained: false
---

### Course information

BIOS/BCB 784

Fall 2024 -- 1:25-2:40 MoWe -- [McGavran 2306](https://sph.unc.edu/room/2306-mcgavran-greenberg-hall/)

This course makes extensive use of R and assumes basic familiarity with
base R (not packages) as a prerequisite. 
A self-quiz is available 
[here](https://github.com/biodatascience/compbio/blob/gh-pages/selfquiz/selfquiz.R), 
with answers provided 
[here](https://github.com/biodatascience/compbio/raw/gh-pages/selfquiz/selfquiz_answers.rda).
You can also find a
[list of base R functions](https://github.com/biodatascience/compbio_src/blob/master/eda/functions.md)
that one should be familiar with.

**For 2024 BCB students**: BCB 720 (background on statistical inference,
i.e. working with the likelihood for parameter inference, conditional
probabilities) is suitable as a pre-requisite for BIOS/BCB 784.

### Schedule and course notes

For `Rmd` files, go to the
[course repo](https://github.com/biodatascience/compbio_src) and navigate the
directories, or best of all to clone the repo and navigate within
RStudio. 

```{r echo=FALSE}
hw <- function(x) paste0("https://github.com/biodatascience/compbio_src/blob/master/",x,".Rmd")
```

| Week | Topic | Dir. | HW | HTML | Title |
|:-----|:------|:-----|:---|:-----|:------|
| | Bio Intro / GitHub | `-` | <!--[HW0](`r hw("github/github_HW")`)--> | [github](github.html) | RStudio, git, and GitHub |
| | Simple EDA | `eda` | | [EDA](eda/EDA.html) | Exploratory data analysis |
| | | | | [NAs](eda/NAs.html) | Missing values in R |
| | | | | [brain RNA](eda/brain_RNA.html) | Exploring brain RNA |
| | Bioconductor I | `bioc` | | [objects](bioc/objects.html) | Bioc data objects |
| | | | | [ranges](bioc/ranges.html) | Genomic ranges |
| | | | |  [GRL](bioc/GRL.html) | GRangesList: lists of ranges |
| | Bioconductor II | | | [anno](bioc/anno.html) | Accessing annotations |
| | | | | [strings](bioc/strings.html) | Manipulating DNA strings |
| | Multiple testing | `test` | | [multtest](test/multtest.html) | FDR and Benjamini-Hochberg |
| | | | | [localfdr](test/localfdr.html) | Local false discovery rate |
| | | | | [IDR](test/IDR.html) | Irreproducible discovery rate |
| | Distances & norm. I | `dist` | | [distances](dist/distances.html) | Distances in high dimensions |
| | | | | [hclust](dist/hclust.html) | Hierarchical clustering |
| | Models and EM | `model` | | [EM](model/EM.html) | Expectation maximization |
| | | | |  [motif](model/motif.html) | EM for finding DNA motifs |
| | ChIP-seq | | | | (Slides on Sakai) |
| | Motifs part II | | | | (In-class EM notes posted to GH) |
| | Distances & norm. II | `dist` | | [batch](dist/batch.html) | Batch effects and sources |
| | | | |  [sva](dist/sva.html) | Surrogate variable analysis |
| | | | | | Batch effect solutions |
| | Hierarchical models | `hier` | | [hierarchical](hier/hierarchical.html) | Hierarchical models |
| | | | |  [jamesstein](hier/jamesstein.html) | James-Stein estimator app |
| | Signal processing | `signal` | | [hmm](signal/hmm.html) | Hidden Markov Models |
| | Tidy genomics | `-` | | [tidy](https://tidyomics.github.io/tidy-ranges-tutorial/) | Tidy ranges tutorial |
| | Network analysis | `net` | | [network](net/network.html) | Network analysis |

### Reading list

* **What is the role of the computational biologist / statistician?**
    - [All biology is computational biology](http://journals.plos.org/plosbiology/article?id=10.1371/journal.pbio.2002050) Florian Markowetz
    - [Questions, Answers and Statistics](https://www.stat.berkeley.edu/~sandrine/Docs/TerrySelectedWorksSpringer/Version1/Nolan/Nolan.pdf) Deborah Nolan
	- [50 Years of Data Science](https://dl.dropboxusercontent.com/u/23421017/50YearsDataScience.pdf) David Donoho
	- [The Future of Data Analysis](https://projecteuclid.org/euclid.aoms/1177704711) John Tukey (this article, discussed by Donoho, is from 1962)
    - [Ten Simple Rules for Effective Statistical Practice](http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004961) Kass, Caffo, Davidian, Meng, Yu, and Reid
    - [Statistical Modeling: The Two Cultures](https://projecteuclid.org/euclid.ss/1009213726) Leo Breiman 
* **Exploratory data analysis**
    - [dplyr - A Grammar of Data Manipulation](http://dplyr.tidyverse.org/)
	- [ggplot2](http://ggplot2.org/)
	- [The Genotype-Tissue Expression (GTEx) pilot analysis: Multitissue gene regulation in humans](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4547484/)
* **Bioconductor**
    - [Orchestrating high-throughput genomic analysis with Bioconductor](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4509590/) Huber et al 
* **Distances and normalization**
    - [Differential expression analysis for sequence count data](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3218662/) Simon Anders and Wolfgang Huber
    - [Tackling the widespread and critical impact of batch effects in high-throughput data](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3880143/) Leek et al
    - [Capturing Heterogeneity in Gene Expression Studies by Surrogate Variable Analysis](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1994707/) Jeffrey Leek and John Storey
	- [Normalization of RNA-seq data using factor analysis of control genes or samples](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4404308/) Risso et al
	- [Using probabilistic estimation of expression residuals (PEER) to obtain increased power and interpretability of gene expression analyses](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3398141/) Stegle et al
* **Multiple testing**
    - [Controlling the False Discovery Rate: A Practical and Powerful Approach to Multiple Testing](http://www.jstor.org/stable/2346101) Yoav Benjamini and Yosef Hochberg
	- [A direct approach to false discovery rates](http://genomics.princeton.edu/storeylab//papers/directfdr.pdf) John Storey
	- [Statistical significance for genomewide studies](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC170937/) John Storey and Robert Tibshirani
	- [Large-scale simultaneous hypothesis testing](http://www.tandfonline.com/doi/abs/10.1198/016214504000000089) Bradley Efron
	- [Empirical Bayes Analysis of a Microarray Experiment](http://dx.doi.org/10.1198/016214501753382129) Efron et al
	- [Measuring reproducibility of high-throughput experiments](https://projecteuclid.org/euclid.aoas/1318514284) Li et al
* **Expectation maximization**
    - [What is the expectation maximization algorithm?](https://www.nature.com/nbt/journal/v26/n8/full/nbt1406.html) Chuong B Do and Serafim Batzoglou
	- [Gaussian mixture models and the EM algorithm](https://people.csail.mit.edu/rameshvs/content/gmm-em.pdf) Ramesh Sridharan
	- [EM algorithm notes](http://cs229.stanford.edu/notes/cs229-notes8.pdf) Andrew Ng
	- [MEME: discovering and analyzing DNA and protein sequence motifs](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1538909/) Bailey et al
* **Hierarchical models**
    - [Linear models and empirical Bayes methods for assessing differential expression in microarray experiments](http://www.statsci.org/smyth/pubs/ebayes.pdf) Gordon Smyth
    - [Analyzing â€™omics data using hierarchical models](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2904972/) Hongkai Ji and X Shirley Liu
	- [Stein's Paradox in Statistics](http://statweb.stanford.edu/~ckirby/brad/other/Article1977.pdf) Bradley Efron and Carl Morris
	- [Stein's estimation rule and its competitors - an empirical Bayes approach](https://www.jstor.org/stable/2284155) Bradley Efron and Carl Morris
* **Signal processing**
    - [An Introduction to Hidden Markov Models](http://ieeexplore.ieee.org/abstract/document/1165342/?reload=true) Lawrence Rabiner and Biing-Hwang Juang
    - [Hidden Markov models approach to the analysis of array CGH data](http://www.sciencedirect.com/science/article/pii/S0047259X04000260) Fridlyand et al
* **Network analysis**
    - [Static And Dynamic DNA Loops Form AP-1 Bound Activation Hubs During Macrophage Development](http://dx.doi.org/10.1016/j.molcel.2017.08.006) Phanstiel et al

### Resources

* [Online R Classes and Resources](http://genomicsclass.github.io/book/pages/resources.html)
* Rafael Irizarry and Michael Love, "Data Analysis for the Life Sciences" 
[Free PDF](https://leanpub.com/dataanalysisforthelifesciences),
[HTML](http://genomicsclass.github.io/book/)
* [Kasper Hansen, "Bioconductor for Genomic Data Science"](https://kasperdanielhansen.github.io/genbioconductor/)
* [Aaron Quinlan, "Applied Computational Genomics" (Slides)](https://github.com/quinlan-lab/applied-computational-genomics)
* [Jennifer Bryan et al, Stat 545](http://stat545.com/)
* [Florian Markowetz, "You Are Not Working for Me; I Am Working with You"](https://doi.org/10.1371/journal.pcbi.1004387)
* [Tips to succeed in Computational Biology research](tips.html)

Some R resources

* [Quick-R](http://www.statmethods.net/)
* [R for Matlab users](http://mathesaurus.sourceforge.net/octave-r.html)
* [R Cookbook](http://www.cookbook-r.com/)
* [R Reference Card](https://cran.r-project.org/doc/contrib/Short-refcard.pdf)
* [Advanced R](http://adv-r.had.co.nz/)
* [Bioconductor Reference Card](https://github.com/mikelove/bioc-refcard)

### Wait, how come ___ is missing?

This is not nearly a complete list of topics in computational
biology. The students taking the course are mostly graduate students
in biostatistics, who have statistical background but not much
exposure to genomic or biological datasets. Classic computational
biology topics, such as alignment algorithms or molecular dynamics,
are not covered, but instead the focus is on exploring genomic
datasets and introducing the key statistical models that flourish in
the high throughput setting (normalization, false discovery rate
calculation, EM algorithm, hierarchical models, HMM, etc.). The course
also focuses on R/Bioconductor, as this is a familiar tool for most
of the students, and allows them to jump in to the data analysis.
The goal is that exposure to these topics and these datasets will
allow them to more effectively read the literature and pursue topics
in biology and biomedical research.

---

This page was last updated on `r format(Sys.time(), "%m/%d/%Y")`. 

* [GitHub repo](https://github.com/biodatascience/compbio_src)
* [License](https://github.com/biodatascience/compbio_src/blob/master/LICENSE)
